jax/scipy/sparse/linalg.py

```diff
@@ -167,10 +188,434 @@ def cg(A, b, x0=None, *, tol=1e-5, atol=0.0, maxiter=None, M=None):
+  H_row. Then constructs and applies a new Givens rotation that eliminates
+  H_row's k'th element.
+  """
+  # This call successively applies each of the
+  # Givens rotations stored in givens[:, :k] to H_col.
+
+  def apply_ith_rotation(i, H_row):
+    cs, sn = givens[i, :]
+    H_i = cs * H_row[i] - sn * H_row[i + 1]
+    H_ip1 = sn * H_row[i] + cs * H_row[i + 1]
+    H_row = H_row.at[i].set(H_i)
+    H_row = H_row.at[i + 1].set(H_ip1)
+    return H_row
+
+  R_row = lax.fori_loop(0, k, apply_ith_rotation, H_row)
+
+  def givens_rotation(v1, v2):
+    t = jnp.sqrt(v1**2 + v2**2)
+    cs = v1 / t
+    sn = -v2 / t
+    return cs, sn
+  givens_factors = givens_rotation(R_row[k], R_row[k + 1])
+  givens = givens.at[k, :].set(givens_factors)
+  cs_k, sn_k = givens_factors
+
+  R_row = R_row.at[k].set(cs_k * R_row[k] - sn_k * R_row[k + 1])
+  R_row = R_row.at[k + 1].set(0.)
+  return R_row, givens
+
+
+def _gmres_qr(A, b, x0, unit_residual, residual_norm, inner_tol, restart, M):
+  """
+  Implements a single restart of GMRES. The restart-dimensional Krylov subspace
+  K(A, x0) = span(A(x0), A@x0, A@A@x0, ..., A^restart @ x0) is built, and the
+  projection of the true solution into this subspace is returned.
+
+  This implementation builds the QR factorization during the Arnoldi process.
+  """
+  # https://www-users.cs.umn.edu/~saad/Calais/PREC.pdf
+  #  residual = _sub(b, A(x0))
+  #  unit_residual, beta = _safe_normalize(residual, return_norm=True)
+
+  V = tree_map(
+      lambda x: jnp.pad(x[..., None], ((0, 0),) * x.ndim + ((0, restart),)),
+      unit_residual,
+  )
+  dtype = jnp.result_type(*tree_leaves(b))
+  R = jnp.eye(restart, restart + 1, dtype=dtype) # eye to avoid constructing
+                                                 # a singular matrix in case
+                                                 # of early termination.
+  b_norm = _norm_tree(b)
+
+  givens = jnp.zeros((restart, 2), dtype=dtype)
+  beta_vec = jnp.zeros((restart + 1), dtype=dtype)
+  beta_vec = beta_vec.at[0].set(residual_norm)
+
+  def loop_cond(carry):
+    k, err, _, _, _, _ = carry
+    return lax.cond(k < restart,
+                    lambda x: x[0] > x[1],
+                    lambda x: False,
+                    (err, inner_tol))
+    # return k < restart and err > tol
+
+  def arnoldi_qr_step(carry):
+    k, _, V, R, beta_vec, givens = carry
+    V, H, _ = kth_arnoldi_iteration(k, A, M, V, R, inner_tol)
+    R_row, givens = apply_givens_rotations(H[k, :], givens, k)
+    R = R.at[k, :].set(R_row[:])
+    cs, sn = givens[k, :] * beta_vec[k]
+    beta_vec = beta_vec.at[k].set(cs)
+    beta_vec = beta_vec.at[k + 1].set(sn)
+    err = jnp.abs(sn) / b_norm
+    return k + 1, err, V, R, beta_vec, givens
+
+  carry = (0, residual_norm, V, R, beta_vec, givens)
+  carry = lax.while_loop(loop_cond, arnoldi_qr_step, carry)
+  k, residual_norm, V, R, beta_vec, _ = carry
+  _ = k # Until we figure out how to pass this to the user.
```

Comment:
Maybe use `del` here for unused variables? That enforces that you can't use `k`.
```suggestion
  del k  # Until we figure out how to pass this to the user.
```